src/hotspot/cpu/aarch64/aarch64_sve.ad

```diff
@@ -3993,3 +3992,89 @@ instruct loadshuffleL(vReg dst, vReg src)
+
+instruct rearrangeB(vReg dst, vReg src, vReg shuffle)
+%{
+  predicate(UseSVE > 0 &&
+            n->bottom_type()->is_vect()->element_basic_type() == T_BYTE);
+  match(Set dst (VectorRearrange src shuffle));
+  ins_cost(SVE_COST);
+  format %{ "sve_tbl $dst, B, $src, $shuffle\t# vector rearrange (B)" %}
+  ins_encode %{
+    __ sve_tbl(as_FloatRegister($dst$$reg), __ B,
+               as_FloatRegister($src$$reg), as_FloatRegister($shuffle$$reg));
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct rearrangeS(vReg dst, vReg src, vReg shuffle)
+%{
+  predicate(UseSVE > 0 &&
+            n->bottom_type()->is_vect()->element_basic_type() == T_SHORT);
+  match(Set dst (VectorRearrange src shuffle));
+  ins_cost(SVE_COST);
+  format %{ "sve_tbl $dst, H, $src, $shuffle\t# vector rearrange (H)" %}
+  ins_encode %{
+    __ sve_tbl(as_FloatRegister($dst$$reg), __ H,
+               as_FloatRegister($src$$reg), as_FloatRegister($shuffle$$reg));
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct rearrangeI(vReg dst, vReg src, vReg shuffle)
+%{
+  predicate(UseSVE > 0 &&
+            n->bottom_type()->is_vect()->element_basic_type() == T_INT);
+  match(Set dst (VectorRearrange src shuffle));
+  ins_cost(SVE_COST);
+  format %{ "sve_tbl $dst, S, $src, $shuffle\t# vector rearrange (S)" %}
+  ins_encode %{
+    __ sve_tbl(as_FloatRegister($dst$$reg), __ S,
+               as_FloatRegister($src$$reg), as_FloatRegister($shuffle$$reg));
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct rearrangeF(vReg dst, vReg src, vReg shuffle)
+%{
+  predicate(UseSVE > 0 &&
+            n->bottom_type()->is_vect()->element_basic_type() == T_FLOAT);
+  match(Set dst (VectorRearrange src shuffle));
+  ins_cost(SVE_COST);
+  format %{ "sve_tbl $dst, S, $src, $shuffle\t# vector rearrange (S)" %}
+  ins_encode %{
+    __ sve_tbl(as_FloatRegister($dst$$reg), __ S,
+               as_FloatRegister($src$$reg), as_FloatRegister($shuffle$$reg));
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct rearrangeL(vReg dst, vReg src, vReg shuffle)
+%{
+  predicate(UseSVE > 0 &&
+            n->bottom_type()->is_vect()->element_basic_type() == T_LONG);
+  match(Set dst (VectorRearrange src shuffle));
+  ins_cost(SVE_COST);
+  format %{ "sve_tbl $dst, D, $src, $shuffle\t# vector rearrange (D)" %}
+  ins_encode %{
+    __ sve_tbl(as_FloatRegister($dst$$reg), __ D,
+               as_FloatRegister($src$$reg), as_FloatRegister($shuffle$$reg));
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct rearrangeD(vReg dst, vReg src, vReg shuffle)
+%{
+  predicate(UseSVE > 0 &&
+            n->bottom_type()->is_vect()->element_basic_type() == T_DOUBLE);
+  match(Set dst (VectorRearrange src shuffle));
+  ins_cost(SVE_COST);
+  format %{ "sve_tbl $dst, D, $src, $shuffle\t# vector rearrange (D)" %}
+  ins_encode %{
+    __ sve_tbl(as_FloatRegister($dst$$reg), __ D,
+               as_FloatRegister($src$$reg), as_FloatRegister($shuffle$$reg));
+  %}
+  ins_pipe(pipe_slow);
+%}
```

Comment:
At least merge `long/double` and `float/int` ? The format looks the same.

Reply:
I think it's my fault. The comment is `L` for `rearrangeL` instead of `D`. I will fix it in my next patch.

Reply:
I also prefer merging them instead of having two separate rules. From the instruction point of view, there's only D, not L.

Reply:
> I also prefer merging them instead of having two separate rules. From the instruction point of view, there's only D, not L.

Thank you for your review.

* I think that the comment is the node's comment (or rule's comment)  instead of single instruction's comment. This rule is `rearrangeL` which means rearranging long type so I think that the comment should be `L` instead of `D`.
* In `aarch64_neon.ad`, the comments are `B``I``F``D` and so on.

Reply:
> > * I think that the comment is the node's comment (or rule's comment)  instead of single instruction's comment. This rule is `rearrangeL` which means rearranging long type so I think that the comment should be `L` instead of `D`.
> > * In `aarch64_neon.ad`, the comments are ` B`I`F`D `` and so on.
> 
> I assume you mean comment in format section. In aarch64_neon, some are S while some are F, but I don't think there's a need to generate two rules just for different format comments. If there's any, I would suggest to fix them as well.

Of course, in some case `L` and `D` can be zipped in one rule. However, in some case, `L` and `D` is difference. `D` is float type.  That is what I mentioned before `the comment is for the whole rule`.  In the comment, we should show the data type. ;-)

Reply:
Some cases are different, but I don't think we need to distinguish floating point data types for this operation as they are actually the same thing and generate the same code.

Reply:
Agree with @nsjian . I prefer to merge all these rules into one as we have discussed too much times before, as there is no instruction difference for different types.  Anyway, we can have a separate patch to unify the styles in ad file.

Reply:
> Some cases are different, but I don't think we need to distinguish floating point data types for this operation as they are actually the same thing and generate the same code.

Yes, I think for this rule `L` and `D` have the same form. However, the comment itself should be consistent with others. In other case, we should distinguish floating point data types. That is why I choose to use `L` instead of `D`.

Reply:
If you really want to have a correct comment, you can comment as something like "(L/D)" - though I doubt if it really make any sense for your testing/debugging.